File: rtl/vc707/host/microgpt_eth.py
#!/usr/bin/env python3
"""UDP host for vc707_microgpt_eth.

The FPGA listens at 192.168.1.42:19783 (UDP). It also responds to ARP
requests for that IP from MAC 02:00:00:4d:47:31 — so once the host can
ARP-resolve the FPGA, ordinary UDP just works.

No raw socket / no special privileges required.

Usage:
  ./microgpt_eth.py probe
  ./microgpt_eth.py gen --steps 15 --temperature 0.5 --seed 2
  ./microgpt_eth.py --fpga 192.168.1.42 --port 19783 probe
"""

import argparse
import socket
import struct
import sys
import time

DEFAULT_FPGA_IP   = "192.168.1.42"
DEFAULT_FPGA_PORT = 19783

FT_REG_WRITE = 0x01
FT_REG_READ  = 0x02
FT_REG_RSP   = 0x03
FT_HEARTBEAT = 0x04
FT_NAK       = 0x05
FT_ACK       = 0x06

# 32-bit register-word offsets — must match vc707_microgpt_eth.sv.
REG_MAGIC       = 0x000
REG_VERSION     = 0x001
REG_CTRL        = 0x002
REG_STATUS      = 0x003
REG_GEN_CFG     = 0x004
REG_SEED        = 0x005
REG_TOP_LOGIT   = 0x006
REG_OUT_BASE    = 0x018
REG_PERF_CYC    = 0x036
REG_LOGITS_BASE = 0x040


def open_socket(timeout: float = 0.5) -> socket.socket:
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.settimeout(timeout)
    return s


def reg_write(sock, peer, writes, seq: int = 0) -> None:
    """writes = [(addr16, data32), ...] (1..16 entries)."""
    assert 1 <= len(writes) <= 16
    # payload byte 0=type, 1=seq, 2=n, 3-7=pad, then n*8 bytes
    body = struct.pack("<BBBBBBBB", FT_REG_WRITE, seq, len(writes), 0, 0, 0, 0, 0)
    for addr, data in writes:
        body += struct.pack("<HIH", addr & 0xFFFF, data & 0xFFFFFFFF, 0)
    sock.sendto(body, peer)


def reg_read(sock, peer, start_addr: int, nwords: int, seq: int = 1):
    """Returns list of nwords ints. Blocks until REG_RSP arrives (or raises)."""
    assert 1 <= nwords <= 19
    # payload byte 0=type, 1=seq, 2-3=start_addr LE, 4=nwords, 5-7=pad
    body = struct.pack("<BBHBBBB", FT_REG_READ, seq,
                       start_addr & 0xFFFF, nwords & 0xFF, 0, 0, 0)
    sock.sendto(body, peer)

    deadline = time.time() + 1.0
    while time.time() < deadline:
        try:
            buf, src = sock.recvfrom(2048)
        except socket.timeout:
            continue
        if len(buf) < 8 or buf[0] != FT_REG_RSP or buf[1] != seq:
            continue
        r_start, r_n = struct.unpack_from("<HB", buf, 2)
        return [struct.unpack_from("<I", buf, 8 + 4 * i)[0] for i in range(r_n)]
    raise TimeoutError(f"no REG_RSP for {nwords} words @ {start_addr:#x}")


def cmd_probe(args, sock, peer):
    vals = reg_read(sock, peer, REG_MAGIC, 4)
    magic_ascii = bytes((vals[0] >> i) & 0xFF for i in (0, 8, 16, 24)).decode("ascii", "replace")
    print(f"  magic   = {vals[0]:#010x}  ({magic_ascii!r})")
    print(f"  version = {vals[1]:#010x}")
    print(f"  ctrl    = {vals[2]:#010x}")
    print(f"  status  = {vals[3]:#010x}")


def cmd_gen(args, sock, peer):
    cfg = ((args.temperature_q88 & 0xFFFF) << 16) | ((args.steps & 0xFF) << 8)
    reg_write(sock, peer, [
        (REG_GEN_CFG, cfg),
        (REG_SEED,    args.seed & 0xFFFFFFFF),
    ], seq=10)
    reg_write(sock, peer, [(REG_CTRL, 0x1)], seq=11)

    t0 = time.time()
    while time.time() - t0 < 5.0:
        s = reg_read(sock, peer, REG_STATUS, 1, seq=12)[0]
        if s & 0x4:
            break
        time.sleep(0.005)
    else:
        raise TimeoutError("generation didn't complete in 5s")

    out_len = (s >> 24) & 0xFF
    print(f"  status   = {s:#010x}  out_len={out_len}")
    tokens = reg_read(sock, peer, REG_OUT_BASE, max(1, out_len), seq=13)
    chars = "abcdefghijklmnopqrstuvwxyz"
    name = "".join(chars[t & 0x1F] for t in tokens if t < 26)
    print(f"  tokens   = {tokens}")
    print(f"  name     = {name!r}")


def cmd_listen(args, sock, peer):
    """Just print HEARTBEAT frames as they arrive."""
    sock.settimeout(2.0)
    print(f"listening for HEARTBEAT from {peer[0]}:{peer[1]} ...")
    try:
        while True:
            try:
                buf, src = sock.recvfrom(2048)
            except socket.timeout:
                continue
            if not buf or buf[0] != FT_HEARTBEAT:
                continue
            state, last_token, out_len, done_flags = struct.unpack_from("BBBB", buf, 2)
            uptime, = struct.unpack_from("<I", buf, 8)
            print(f"hb: state={state} last_token={last_token} out_len={out_len} "
                  f"flags={done_flags:#04x} uptime={uptime}s")
    except KeyboardInterrupt:
        print()


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--fpga", default=DEFAULT_FPGA_IP, help=f"FPGA IP (default {DEFAULT_FPGA_IP})")
    p.add_argument("--port", type=int, default=DEFAULT_FPGA_PORT,
                   help=f"FPGA UDP port (default {DEFAULT_FPGA_PORT})")
    sub = p.add_subparsers(dest="cmd", required=True)

    sub.add_parser("probe")
    sub.add_parser("listen")

    g = sub.add_parser("gen")
    g.add_argument("--steps", type=int, default=15)
    g.add_argument("--temperature", type=float, default=0.5)
    g.add_argument("--seed", type=lambda x: int(x, 0), default=2)

    args = p.parse_args()
    if hasattr(args, "temperature"):
        args.temperature_q88 = max(1, min(0xFFFF, int(round(args.temperature * 256))))

    sock = open_socket()
    peer = (args.fpga, args.port)

    if args.cmd == "probe":   cmd_probe(args, sock, peer)
    elif args.cmd == "gen":   cmd_gen(args, sock, peer)
    elif args.cmd == "listen": cmd_listen(args, sock, peer)


if __name__ == "__main__":
    main()

File: rtl/vc707/host/microgpt_eth.c
/* microgpt_eth.c — minimal raw-Ethernet launcher for vc707_microgpt_eth.
 *
 * Linux only (AF_PACKET). Build:
 *   gcc -O2 -Wall -Wextra microgpt_eth.c -o microgpt_eth
 * Run (needs CAP_NET_RAW):
 *   sudo ./microgpt_eth eth0 probe
 *   sudo ./microgpt_eth eth0 gen 15 0.5 2
 *
 * Mirrors microgpt_eth.py for callers that want a C path instead.
 */

#define _GNU_SOURCE
#include <arpa/inet.h>
#include <errno.h>
#include <linux/if_packet.h>
#include <net/ethernet.h>
#include <net/if.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <unistd.h>

#define ETH_TYPE_MG  0x4D47
#define FT_REG_WRITE 0x01
#define FT_REG_READ  0x02
#define FT_REG_RSP   0x03
#define FT_NAK       0x05
#define FT_ACK       0x06

static const uint8_t FPGA_MAC[6] = {0x02,0x00,0x00,0x4D,0x47,0x31};

#define REG_MAGIC       0x000
#define REG_VERSION     0x001
#define REG_CTRL        0x002
#define REG_STATUS      0x003
#define REG_GEN_CFG     0x004
#define REG_SEED        0x005
#define REG_OUT_BASE    0x018

static int open_iface(const char *iface, uint8_t host_mac[6], int *ifindex)
{
    int s = socket(AF_PACKET, SOCK_RAW, htons(ETH_TYPE_MG));
    if (s < 0) { perror("socket"); return -1; }

    struct ifreq ifr = {0};
    strncpy(ifr.ifr_name, iface, IFNAMSIZ - 1);
    if (ioctl(s, SIOCGIFINDEX, &ifr) < 0) { perror("SIOCGIFINDEX"); close(s); return -1; }
    *ifindex = ifr.ifr_ifindex;

    if (ioctl(s, SIOCGIFHWADDR, &ifr) < 0) { perror("SIOCGIFHWADDR"); close(s); return -1; }
    memcpy(host_mac, ifr.ifr_hwaddr.sa_data, 6);

    struct sockaddr_ll sll = {0};
    sll.sll_family   = AF_PACKET;
    sll.sll_protocol = htons(ETH_TYPE_MG);
    sll.sll_ifindex  = *ifindex;
    if (bind(s, (struct sockaddr *)&sll, sizeof(sll)) < 0) {
        perror("bind"); close(s); return -1;
    }

    struct timeval tv = { .tv_sec = 0, .tv_usec = 500000 };
    setsockopt(s, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
    return s;
}

static int send_frame(int s, int ifindex, const uint8_t host_mac[6],
                      const uint8_t *body, size_t blen)
{
    uint8_t buf[64 + 256];
    size_t off = 0;
    memcpy(buf + off, FPGA_MAC, 6); off += 6;
    memcpy(buf + off, host_mac, 6); off += 6;
    buf[off++] = (ETH_TYPE_MG >> 8) & 0xFF;
    buf[off++] = ETH_TYPE_MG & 0xFF;
    memcpy(buf + off, body, blen); off += blen;
    while (off < 60) buf[off++] = 0;

    struct sockaddr_ll sll = {0};
    sll.sll_family   = AF_PACKET;
    sll.sll_protocol = htons(ETH_TYPE_MG);
    sll.sll_ifindex  = ifindex;
    sll.sll_halen    = 6;
    memcpy(sll.sll_addr, FPGA_MAC, 6);

    return sendto(s, buf, off, 0, (struct sockaddr *)&sll, sizeof(sll));
}

static int reg_write(int s, int ifindex, const uint8_t host_mac[6], uint8_t seq,
                     const uint16_t *addrs, const uint32_t *data, uint8_t n)
{
    uint8_t body[2 + 8 + 16 * 8];
    size_t off = 0;
    body[off++] = FT_REG_WRITE; body[off++] = seq;
    body[off++] = n; body[off++] = seq;
    memset(body + off, 0, 6); off += 6;
    for (uint8_t i = 0; i < n; i++) {
        body[off++] =  addrs[i]        & 0xFF;
        body[off++] = (addrs[i] >> 8)  & 0xFF;
        body[off++] =  data[i]         & 0xFF;
        body[off++] = (data[i]  >> 8)  & 0xFF;
        body[off++] = (data[i]  >> 16) & 0xFF;
        body[off++] = (data[i]  >> 24) & 0xFF;
        body[off++] = 0; body[off++] = 0;
    }
    return send_frame(s, ifindex, host_mac, body, off);
}

static int reg_read(int s, int ifindex, const uint8_t host_mac[6], uint8_t seq,
                    uint16_t start_addr, uint8_t nwords, uint32_t *out)
{
    uint8_t body[2 + 8] = {FT_REG_READ, seq,
                            start_addr & 0xFF, (start_addr >> 8) & 0xFF,
                            nwords, seq, 0, 0, 0, 0};
    if (send_frame(s, ifindex, host_mac, body, sizeof(body)) < 0)
        return -1;

    uint8_t buf[2048];
    for (int tries = 0; tries < 10; tries++) {
        ssize_t n = recv(s, buf, sizeof(buf), 0);
        if (n < 24) continue;
        if (buf[12] != ((ETH_TYPE_MG >> 8) & 0xFF) || buf[13] != (ETH_TYPE_MG & 0xFF))
            continue;
        if (buf[14] == FT_NAK) {
            fprintf(stderr, "NAK seq=%u err=0x%02x\n", buf[16], buf[17]);
            return -1;
        }
        if (buf[14] != FT_REG_RSP || buf[15] != seq) continue;
        uint8_t r_n = buf[18];
        for (uint8_t i = 0; i < r_n && i < nwords; i++) {
            out[i] = (uint32_t)buf[24 + 4*i]
                   | ((uint32_t)buf[25 + 4*i] << 8)
                   | ((uint32_t)buf[26 + 4*i] << 16)
                   | ((uint32_t)buf[27 + 4*i] << 24);
        }
        return r_n;
    }
    return -1;
}

static int cmd_probe(int s, int ifindex, const uint8_t host_mac[6])
{
    uint32_t v[4];
    if (reg_read(s, ifindex, host_mac, 1, REG_MAGIC, 4, v) != 4) {
        fprintf(stderr, "probe failed\n"); return 1;
    }
    printf("magic   = 0x%08x  ('%c%c%c%c')\n", v[0],
        v[0] & 0xFF, (v[0]>>8)&0xFF, (v[0]>>16)&0xFF, (v[0]>>24)&0xFF);
    printf("version = 0x%08x\n", v[1]);
    printf("ctrl    = 0x%08x\n", v[2]);
    printf("status  = 0x%08x\n", v[3]);
    return 0;
}

static int cmd_gen(int s, int ifindex, const uint8_t host_mac[6],
                   int steps, double temperature, uint32_t seed)
{
    uint16_t temp_q88 = (uint16_t)(temperature * 256.0);
    uint32_t cfg = ((uint32_t)temp_q88 << 16) | ((uint32_t)(steps & 0xFF) << 8);
    uint16_t a[2] = {REG_GEN_CFG, REG_SEED};
    uint32_t d[2] = {cfg, seed};
    if (reg_write(s, ifindex, host_mac, 10, a, d, 2) < 0) {
        perror("reg_write cfg"); return 1;
    }
    uint16_t a2[1] = {REG_CTRL};
    uint32_t d2[1] = {0x1};
    if (reg_write(s, ifindex, host_mac, 11, a2, d2, 1) < 0) {
        perror("reg_write start"); return 1;
    }

    uint32_t status = 0;
    for (int i = 0; i < 1000; i++) {
        if (reg_read(s, ifindex, host_mac, 12, REG_STATUS, 1, &status) != 1)
            continue;
        if (status & 0x4) break;
        usleep(5000);
    }
    if (!(status & 0x4)) { fprintf(stderr, "timeout, status=0x%08x\n", status); return 1; }

    uint8_t out_len = (status >> 24) & 0xFF;
    if (out_len == 0) out_len = 1;
    uint32_t toks[16];
    int n = reg_read(s, ifindex, host_mac, 13, REG_OUT_BASE, out_len, toks);
    if (n <= 0) { fprintf(stderr, "token read failed\n"); return 1; }

    printf("status=0x%08x out_len=%u tokens=[", status, out_len);
    for (int i = 0; i < n; i++) printf(i ? ",%u" : "%u", toks[i] & 0xFF);
    printf("] name=\"");
    for (int i = 0; i < n; i++) {
        uint8_t t = toks[i] & 0xFF;
        if (t < 26) putchar('a' + t);
    }
    printf("\"\n");
    return 0;
}

int main(int argc, char **argv)
{
    if (argc < 3) {
        fprintf(stderr, "usage: %s <iface> probe\n"
                        "       %s <iface> gen <steps> <temperature> <seed>\n",
                argv[0], argv[0]);
        return 2;
    }
    uint8_t host_mac[6]; int ifindex;
    int s = open_iface(argv[1], host_mac, &ifindex);
    if (s < 0) return 1;

    int rc = 0;
    if (!strcmp(argv[2], "probe")) {
        rc = cmd_probe(s, ifindex, host_mac);
    } else if (!strcmp(argv[2], "gen") && argc >= 6) {
        rc = cmd_gen(s, ifindex, host_mac,
                     atoi(argv[3]), atof(argv[4]), strtoul(argv[5], NULL, 0));
    } else {
        fprintf(stderr, "unknown command\n"); rc = 2;
    }
    close(s);
    return rc;
}
